### PR TITLE
Purecn/intervalfile

### LINKED ISSUE #2937: new module: purecn/intervalfile
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

### PR DESCRIPTION
**Pull request by lbeltrame**
## PR checklist

Initial implementation of the purcn/intervalfile module. 

Closes #2937  

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

**Comment by lbeltrame** [comment]
The module now passes the tests locally. The blocker issues / PRs need to be fixed / merged before this can go out of draft, though.

**Comment by lbeltrame** [comment]
Blocked by https://github.com/BioContainers/multi-package-containers/pull/2554.

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/main.nf:32]:
```suggestion
    Rscript "\$library_path"/PureCN/extdata/IntervalFile.R \\
        --in-file ${target_bed} \\
```

We try to have each flag on a separate line for clarity

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/main.nf:43]: Can you add a stub section? [docs](https://www.nextflow.io/docs/latest/process.html#stub)

File: modules/nf-core/purecn/intervalfile/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/meta.yml:4]: Can you add some more keywords?

File: modules/nf-core/purecn/intervalfile/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/meta.yml:31]:
```suggestion
  - txt:
      type: file
      pattern: "*.txt"
      description: |
        Annotated targets optimized for copy number calling
```
Can you also add `bed` with a description of how to generate the file using the tool?

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/main.nf:34]:
```suggestion
        --out-file ${prefix}.txt \\
```
Do not set hardcoded file names. By setting prefix as the filename, the users have full control over how the output file will be named

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (lbeltrame) [reply modules/nf-core/purecn/intervalfile/main.nf:1]: Do these apply even if the command *has* a `--version` switch? PureCN has one, but it exits with status code 1, and that makes the whole process fails. No problem in doing that, but just asking.

Comment (nvnieuwk) [reply modules/nf-core/purecn/intervalfile/main.nf:1]: That's a weird way of defining the versions flag :). I don't know if there is any way you can circumvent this exit code for the versions flag (maybe using a bash function). Otherwise you can leave it as it is now indeed

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/main.nf:49]: The versions in stub should be created like it would in the normal process:
```suggestion
    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        purecn: ${VERSION}
    END_VERSIONS
```

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [reply modules/nf-core/purecn/intervalfile/main.nf:1]: Can you adjust the manual version according to the guidelines which I stated in the first comment of this thread?

File: modules/nf-core/purecn/intervalfile/meta.yml
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/meta.yml:41]: Meta should also be added to the output in the meta.yml

File: modules/nf-core/purecn/intervalfile/main.nf
Comment (nvnieuwk) [inline modules/nf-core/purecn/intervalfile/main.nf:48]:
```suggestion
    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    def bed = args.contains("--export") ? "touch ${prefix}.bed" : ""
    """
    touch ${prefix}.txt
    ${bed}
```

The `args` and `prefix` are only set in the `script` clause (which isn't run when using `-stub`). For that reason you'll need to specify the `args` and `prefix` again in the `stub` clause. Also it's better to also account for optional file creation if possible (it's simple in this case, if `--export` is in the `args`, you create the BED file)

File: modules/nf-core/purecn/intervalfile/meta.yml
Comment (lbeltrame) [reply modules/nf-core/purecn/intervalfile/meta.yml:41]: Fixed as well (I think).

**Comment by lbeltrame** [comment]
Should `VERSION` stay inside or outside `process`? It's the only last doubt I have about this.

**Comment by nvnieuwk** [comment]
> Should `VERSION` stay inside or outside `process`? It's the only last doubt I have about this.

We prefer inside of the `script` and `stub` clause with a comment next to it. This way you'll notice it better when updating the module. A comment should also be added at the containers/conda to note the developers to manually update the VERSION if they update the containers/conda

**Comment by lbeltrame** [comment]
Ok, the comment for containers is already there, I'll adjust `VERSION` and hopefully it should be it (fingers crossed!)

**Review by nvnieuwk — APPROVED** [review]
Awesome! Good job!

**Comment by lbeltrame** [comment]
I reviewed the guidelines and I'm not sure what's the next step. Wait for someone to merge this? Add this to the merge queue?

**Comment by nvnieuwk** [comment]
You may always press the merge button when you get an approval, I'll do it for you this time :)

**Pull request merged**
